iotdb/src/main/java/org/apache/iotdb/db/engine/filenode/FileNodeProcessor.java

```diff
@@ -1902,14 +1924,43 @@ private FileNodeProcessorStore readStoreFromDisk() throws FileNodeProcessorExcep
                 new IntervalFileNode(OverflowChangeType.NO_CHANGE, null),
                 new ArrayList<IntervalFileNode>(), FileNodeProcessorStatus.NONE, 0));
       } catch (IOException e) {
-        e.printStackTrace();
         throw new FileNodeProcessorException(e);
       }
       return processorStore;
     }
   }
 
-  public String getFileNodeRestoreFilePath() {
-    return fileNodeRestoreFilePath;
-  }
+    public String getFileNodeRestoreFilePath() {
+        return fileNodeRestoreFilePath;
+    }
+
+    /**
+     * Delete data whose timestamp <= 'timestamp' and belong to timeseries deviceId.measurementId.
+     * @param deviceId the deviceId of the timeseries to be deleted.
+     * @param measurementId the measurementId of the timeseries to be deleted.
+     * @param timestamp the delete range is (0, timestamp].
+     */
+    public void delete(String deviceId, String measurementId, long timestamp) throws IOException {
+      // TODO: how to avoid partial deletion?
+      long version = versionController.nextVersion();
+
+      String fullPath = deviceId +
+              IoTDBConstant.PATH_SEPARATOR + measurementId;
+      Deletion deletion = new Deletion(fullPath, version, timestamp);
+      if (currentIntervalFileNode != null && currentIntervalFileNode.containsDevice(deviceId)) {
+        currentIntervalFileNode.getModFile().write(deletion);
+      }
+      for (IntervalFileNode fileNode : newFileNodes) {
```

Comment:
write Deletion on each TsFile? (how about checking whether the startTime of the device in a TsFile is less than the given timestamp?)

Reply:
Added.